clients/python-wrapper/lakefs/object.py

```diff
@@ -357,62 +303,95 @@ def _get_range_string(start, read_bytes=None):
-        super().__init__(repository, reference, path, client=client)
+    _BUFFER_SIZE = 32 * 1024 * 1024
+    _fd: Union[io.BufferedWriter, io.TextIOWrapper]
+    _obj_stats: ObjectStats = None
+
+    def __init__(self,
+                 obj: StoredObject,
+                 mode: WriteModes,
+                 pre_sign: Optional[bool] = None,
+                 content_type: Optional[str] = None,
+                 metadata: Optional[dict[str, str]] = None,
+                 client: Optional[Client] = DEFAULT_CLIENT) -> None:
+
+        if 'x' in mode and obj.exists():  # Requires explicit create
+            raise ObjectExistsException
 
-    def upload(self,
-               data: str | bytes,
-               mode: WriteModes = 'wb',
-               pre_sign: Optional[bool] = None,
-               content_type: Optional[str] = None,
-               metadata: Optional[dict[str, str]] = None) -> WriteableObject:
-        """
-        Creates a new object or overwrites an existing object
-        :param data: The contents of the object to write (can be bytes or string)
-        :param mode: Write mode:
-            'x'     - Open for exclusive creation
-            'xb'    - Open for exclusive creation in binary mode
-            'w'     - Create a new object or truncate if exists
-            'wb'    - Create or truncate in binary mode
-        :param pre_sign: (Optional) Explicitly state whether to use pre_sign mode when uploading the object.
-        If None, will be taken from pre_sign property.
-        :param content_type: (Optional) Explicitly set the object Content-Type
-        :param metadata: (Optional) User metadata
-        :return: The Stat object representing the newly created object
-        :raises
-            ObjectExistsException if object exists and mode is exclusive ('x')
-            ObjectNotFoundException if repo id, reference id or object path does not exist
-            PermissionException if user is not authorized to perform this operation, or operation is forbidden
-            ServerException for any other errors
-        """
         if mode not in get_args(WriteModes):
             raise ValueError(f"invalid write mode: '{mode}'. WriteModes: {WriteModes}")
 
-        if 'x' in mode and self.exists():  # Requires explicit create
-            raise ObjectExistsException
+        self.content_type = content_type
+        self.metadata = metadata
 
-        binary_mode = 'b' in mode
-        if binary_mode and isinstance(data, str):
-            content = data.encode('utf-8')
-        elif not binary_mode and isinstance(data, bytes):
-            content = data.decode('utf-8')
+        open_kwargs = {
+            "encoding": "utf-8" if 'b' not in mode else None,
+            "mode": 'wb+' if 'b' in mode else 'w+',
+            "buffering": self._BUFFER_SIZE,
+        }
+        self._fd = tempfile.TemporaryFile(**open_kwargs)
```

Comment:
so atm we always write the whole thing to the temp file first?

Reply:
Addressing this [comment](https://github.com/treeverse/lakeFS/issues/7033#issuecomment-1824993264), data will be written to the buffer and will not be written to the file as long as the buffer is not filled.